lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+      freeClassicals[op] = SetVector<ShadowDependencyNode *>();
+      ifStack.push_back(op);
+      auto then_block = visitBlock(ifop.getThenEntryBlock(), dependencies);
+      auto else_block = visitBlock(ifop.getElseEntryBlock(), dependencies);
+      if (!then_block || !else_block)
+        return nullptr;
+      ifStack.pop_back();
+
+      SetVector<ShadowDependencyNode *> freeIn = freeClassicals[op];
+      freeClassicals.erase(freeClassicals.find(op));
+
+      newNode = new IfDependencyNode(ifop, dependencies, then_block, else_block,
+                                     freeIn);
+    } else if (isTerminator) {
+      newNode = new TerminatorDependencyNode(op, dependencies);
+    } else {
+      newNode = new OpDependencyNode(op, dependencies);
+    }
+
+    // Dnodeid is the next slot of the dnode vector
+    auto id = perOp.size();
+
+    // Add dnodeid attribute
+    OpBuilder builder(op);
+    op->setAttr("dnodeid", builder.getUI32IntegerAttr(id));
+    perOp.push_back(newNode);
+
+    return newNode;
+  }
+
+  /// Returns the dependency node for the defining operation of \p v
+  /// Assumption: defining operation for \p v exists and already has been
+  /// visited.
+  ///
+  /// If \p v a classical value from a different scope, allocates a
+  /// ShadowDependencyNode, and adds it to the frontier of the parent node
+  /// of the use of \p v that is at the same scope as \p v
+  DependencyNode::DependencyEdge visitValue(Value v) {
+    if (auto barg = dyn_cast<BlockArgument>(v))
+      return DependencyNode::DependencyEdge{argMap[barg], 0};
+
+    auto defOp = v.getDefiningOp();
+    assert(defOp &&
+           "Cannot handle value that is neither a BlockArgument nor OpResult");
+
+    auto resultidx = dyn_cast<OpResult>(v).getResultNumber();
+
+    // Since we walk forward through the ast, every value should be defined
+    // before it is used, so we should have already visited defOp,
+    // and thus should have a memoized dnode for defOp, fail if not
+    assert(defOp->hasAttr("dnodeid") && "No dnodeid found for operation");
+
+    auto id = defOp->getAttr("dnodeid").cast<IntegerAttr>().getUInt();
+    auto dnode = perOp[id];
+
+    if (!ifStack.empty() && defOp->getParentOp() != ifStack.back() &&
+        dnode->isQuantumDependent()) {
+      auto opdnode = static_cast<OpDependencyNode *>(dnode);
+      auto shadow_node = new ShadowDependencyNode{opdnode, resultidx};
+
+      auto parent = ifStack.back();
+
+      while (parent->getParentOp() != defOp->getParentOp())
+        parent = parent->getParentOp();
+
+      freeClassicals[parent].insert(shadow_node);
+
+      return DependencyNode::DependencyEdge{shadow_node, resultidx};
+    }
+
+    return DependencyNode::DependencyEdge{dnode, resultidx};
+  }
+
+  /// Cleans up semi-constructed dependency graph when backing out of running
+  /// DependencyAnalysis because of an encountered error
+  void clean() {
+    // TODO: clean up nodes
```

Comment:
Is this leaking memory?

Reply:
Yeah, a bunch of `DependencyNode`s are getting allocated when building up the `DependencyGraph`, but if there's an illegal operation that causes the analysis to back out, those allocated `DependencyNode`s aren't freed currently. That's why I added the `clean()` function, but I didn't have time to write it. I think it will take an extra data structure to keep track of all the allocated `DependencyNode`s, but also remove the ones inside `IfDependencyNode`s, because the `IfDependencyNode` is responsible for cleaning those up. The tricky part is that some nodes could be inside an `if`, but if there's an illegal operation later in the `if`, then no `IfDependencyNode` would be created and the nodes would have to be explicitly cleaned up by the `DependencyAnalysisEngine`.

Reply:
Thinking about this a bit more, I think it may be as simple as collecting the set of nodes in `visitBlock` and then cleaning up there if an illegal operation is encountered.

Reply:
Thanks, Adam.

Reply:
Will create a separate issue for this one (or add to a new grouped issue).

Reply:
https://github.com/NVIDIA/cuda-quantum/issues/2181 written. Resolving this conversation.